File: src/main/java/net/greenjab/fixedminecraft/registry/block/NewAmethystBlock.java
package net.greenjab.fixedminecraft.registry.block;

import com.mojang.serialization.MapCodec;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.util.RandomSource;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.AmethystBlock;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.RedstoneTorchBlock;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.gameevent.GameEvent;
import net.minecraft.world.level.gameevent.vibrations.VibrationSystem;
import net.minecraft.world.level.redstone.Orientation;
import org.jetbrains.annotations.Nullable;
import org.jspecify.annotations.NonNull;

public class NewAmethystBlock extends AmethystBlock {
    public static final BooleanProperty LIT  = RedstoneTorchBlock.LIT;
    public static final MapCodec<NewAmethystBlock> CODEC = simpleCodec(NewAmethystBlock::new);
    public static final int[] RESONATION_NOTE_PITCHES = {0, 0, 2, 4, 6, 7, 9, 10, 12, 14, 16, 18, 19, 21, 22, 24};

    public NewAmethystBlock(Properties settings) {
        super(settings);
    }

    @Override
    public @NonNull MapCodec<NewAmethystBlock> codec()  {
        return CODEC;
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        return defaultBlockState().setValue(LIT, ctx.getLevel().hasNeighborSignal(ctx.getClickedPos()));
    }

    @Override
    protected void neighborChanged(
            @NonNull BlockState state,
            @NonNull Level level,
            @NonNull BlockPos pos,
            @NonNull Block sourceBlock,
            @Nullable Orientation wireOrientation,
            boolean notify
    ) {
        if (!level.isClientSide()) {
            boolean bl = state.getValue(LIT);
            if (bl != level.hasNeighborSignal(pos)) {
                if (bl) {
                    level.scheduleTick(pos, this, 2);
                } else {
                    level.setBlock(pos, state.cycle(LIT), UPDATE_CLIENTS);
                    int frequency = level.getBestNeighborSignal(pos);
                    level.gameEvent(VibrationSystem.getResonanceEventByFrequency(frequency), pos, GameEvent.Context.of(null, state));
                    int f = RESONATION_NOTE_PITCHES[frequency];
                    level.playSound(null, pos, SoundEvents.AMETHYST_BLOCK_RESONATE, SoundSource.BLOCKS, 1.0f, f);
                    if (!level.getBlockState(pos.above()).isFaceSturdy(level, pos.above(), Direction.DOWN)) ((ServerLevel) level).sendParticles(ParticleTypes.NOTE, pos.getX() + 0.5, pos.getY() + 1, pos.getZ() + 0.5, 0, 1, 0.0, 0.0, frequency / 15.0);
                }
            }
        }
    }

    @Override
    public void tick(BlockState state, @NonNull ServerLevel world, @NonNull BlockPos pos, @NonNull RandomSource random) {
        if (state.getValue(LIT) && !world.hasNeighborSignal(pos)) {
            world.setBlock(pos, state.cycle(LIT), UPDATE_CLIENTS);
        }
    }

    @Override
    public void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(LIT);
    }
}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/block/NewAmethystBlock.kt
package net.greenjab.fixedminecraft.registry.block

import com.mojang.serialization.MapCodec
import net.minecraft.block.AmethystBlock
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.NoteBlock
import net.minecraft.block.RedstoneTorchBlock
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemPlacementContext
import net.minecraft.server.world.ServerWorld
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvents
import net.minecraft.state.StateManager
import net.minecraft.state.property.BooleanProperty
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.random.Random
import net.minecraft.world.World
import net.minecraft.world.block.WireOrientation
import net.minecraft.world.event.GameEvent
import net.minecraft.world.event.Vibrations

class NewAmethystBlock(settings: Settings?) : AmethystBlock(settings) {
    override fun getCodec(): MapCodec<NewAmethystBlock> {
        return CODEC
    }

    override fun getPlacementState(ctx: ItemPlacementContext): BlockState {
        return defaultState.with(LIT, ctx.world.isReceivingRedstonePower(ctx.blockPos)) as BlockState
    }

    override fun neighborUpdate(
        state: BlockState?,
        world: World?,
        pos: BlockPos?,
        sourceBlock: Block?,
        wireOrientation: WireOrientation?,
        notify: Boolean
    ) {
        if (world != null) {
            if (!world.isClient) {
                val bl = state?.get(LIT) as Boolean
                if (bl != world.isReceivingRedstonePower(pos)) {
                    if (bl) {
                        world.scheduleBlockTick(pos, this, 4)
                    } else {
                        world.setBlockState(pos, state.cycle(LIT) as BlockState, NOTIFY_LISTENERS)
                        val frequency = world.getReceivedRedstonePower(pos)
                        world.emitGameEvent(Vibrations.getResonation(frequency), pos, GameEvent.Emitter.of(null, state))
                        val f = RESONATION_NOTE_PITCHES[frequency]
                        world.playSound(null as PlayerEntity?,pos,SoundEvents.BLOCK_AMETHYST_BLOCK_RESONATE,SoundCategory.BLOCKS,1.0f,f)
                    }
                }
            }
        }
    }

    override fun scheduledTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
        if (state.get(LIT) as Boolean && !world.isReceivingRedstonePower(pos)) {
            world.setBlockState(pos, state.cycle(LIT) as BlockState, NOTIFY_LISTENERS)
        }
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        builder.add(LIT)
    }

    init {
        this.defaultState = defaultState.with(LIT, false) as BlockState
    }

    companion object {
        val CODEC: MapCodec<NewAmethystBlock> = createCodec { settings: Settings? ->
            NewAmethystBlock(
                settings
            )
        }
        val LIT: BooleanProperty = RedstoneTorchBlock.LIT

        val RESONATION_NOTE_PITCHES: FloatArray =
            net.minecraft.util.Util.make(FloatArray(16)) { frequency: FloatArray? ->
                val freq: IntArray = intArrayOf(0, 0, 2, 4, 6, 7, 9, 10, 12, 14, 15, 18, 19, 21, 22, 24)
                for (i in 0..15) {
                    frequency?.set(i, NoteBlock.getNotePitch(freq[i]))
                }
            }
    }

}
